src/main.rs

```diff
@@ -71,261 +64,57 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
-            // Record metrics
-            // //////////////////////////////////////////////////////////////
-            let metrics = discv5.metrics();
-            let write_query = WriteQuery::new(
-                Local::now().into(),
-                format!(
-                    "discv5-testground_{}_{}",
-                    run_parameters.test_case, run_parameters.test_run
-                ),
-            )
-            .add_field("active_sessions", metrics.active_sessions as u64)
-            .add_field(
-                "unsolicited_requests_per_second",
-                metrics.unsolicited_requests_per_second as f64,
-            )
-            .add_field("bytes_sent", metrics.bytes_sent as u64)
-            .add_field("bytes_recv", metrics.bytes_recv as u64)
-            .add_tag("instance_seq", instance_info.seq);
-            client.record_metric(write_query).await?;
+    match run_parameters.test_case.clone().as_str() {
+        "find-node" => find_node::find_node(client.clone(), run_parameters.clone()).await?,
+        "eclipse-attack-monopolizing-by-incoming-nodes" => {
+            eclipse::MonopolizingByIncomingNodes::new(run_parameters.clone())
+                .run(client.clone())
+                .await?
         }
-    }
-
-    client
-        .signal_and_wait(
-            STATE_COMPLETED_TO_RUN_FIND_NODE_QUERY,
-            run_parameters.test_instance_count,
-        )
-        .await?;
-
-    // //////////////////////////////////////////////////////////////
-    // Record result of this test and shutdown Discovery v5 server
-    // //////////////////////////////////////////////////////////////
-    if failed {
-        client
-            .record_failure("Failures have happened, please check error logs for details.")
-            .await?;
-    } else {
-        client.record_success().await?;
-    }
-    discv5.shutdown();
+        _ => unreachable!(),
+    };
 
     Ok(())
 }
 
-#[derive(Debug, Serialize, Deserialize)]
-struct InstanceInfo {
-    // The sequence number of this test instance within the test.
-    seq: u64,
-    enr: Enr<CombinedKey>,
-    is_bootstrap_node: bool,
-}
-
-impl InstanceInfo {
-    async fn new(
-        client: &Client,
-        run_parameters: &RunParameters,
-        enr: Enr<CombinedKey>,
-    ) -> Result<Self, Box<dyn std::error::Error>> {
-        let seq = get_instance_seq(client, run_parameters).await?;
-
-        // NOTE: For now, #1 is bootstrap node.
-        let is_bootstrap_node = seq == 1;
-
-        Ok(InstanceInfo {
-            seq,
-            enr,
-            is_bootstrap_node,
-        })
-    }
+// Returns the sequence number of this test instance within the test.
+async fn get_instance_seq(client: &Client) -> Result<u64, testground::errors::Error> {
+    client.signal("get_instance_seq").await
 }
 
-// Returns the sequence number of this test instance within the test.
-async fn get_instance_seq(
+async fn get_group_seq(
```

Comment:
Cool! Would be good to upstream this eventually. For the record: https://github.com/testground/sdk-rust/pull/25#discussion_r900044233

Reply:
Not yet sure how best to expose global and group sequence number. Curious what you think @ackintosh.

Reply:
Yeah, I was thinking of what clean way to do so. 🤔  How about something `Context` object like this in go-sdk?

https://github.com/testground/sdk-go/blob/044f6f8ec366a38f067c167cf7c55546bf60c572/run/init_context.go#L33-L40

---

```rust
// ///////////////////
// 1. `Context` object wraps all of these.
// ///////////////////
// struct Context {
//     client: Client,
//     run_parameter: RunParameters,
//     global_seq: u64,
//     group_seq: u64,
// }
let context = Context::init();
// ctx.client.record_message("");
// ctx.run_parameter
// ctx.global_seq
// ctx.group_seq


// ///////////////////
// 2.`Context` object wraps only static information about the runtime.
// ///////////////////
// struct Context {
//     run_parameter: RunParameters,
//     global_seq: u64,
//     group_seq: u64,
// }
let (client, context) = Client::new();
```

I prefer the 2nd idea (`2. Context object wraps only static information about the runtime`) since the role of the `Context` object is clear. (and if `Context` has `Client`, it seems some ownership issue to occur.)

Reply:
That makes sense. 💡 Your idea looks good and intuitive to me as `Client::new` returns a `Client`, not a tuple.

Reply:
If it's fine with you, I would create a pull request. 😃

Reply:
@ackintosh much appreciated! Thanks.

Reply:
:octocat:  https://github.com/testground/sdk-rust/pull/29